src/lib/crypto/kimchi_backend/gadgets/and.ml

```diff
@@ -0,0 +1,148 @@
+open Core_kernel
+
+open Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint
+
+(* AND *)
+
+(* Boolean And of length bits *)
+let band (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (input1 : Circuit.Field.t) (input2 : Circuit.Field.t) (length : int)
+    (len_xor : int) : Circuit.Field.t =
+  let open Circuit in
+  let output_and =
+    exists Field.typ ~compute:(fun () ->
+        (* Recursively build And gadget with leading Xors and a final Generic gate *)
+        (* It will also check the right lengths of the inputs, no need to do it again *)
+        let xor_output =
```

Comment:
Things in `exists` happens at both circuit creation time and prover time.  I don't think we should call gadget creation code inside an `exists`, otherwise we'd get nested `exists` may perform circuit creation code at prover time.  I think you should move this outside the `exists`.    I'm not sure, so let's see what @mrmr1993 says.